application-collabora-default/src/main/java/com/xwiki/collabora/internal/rest/DefaultWopi.java

```diff
@@ -184,7 +184,7 @@ public Token getToken(String fileId) throws XWikiRestException
 
             Token token = (new ObjectFactory()).createToken();
             token.setUrlSrc(urlSrc);
-            token.setValue(fileTokenManager.getToken(xcontext.getUserReference(), fileId).toString());
+            token.setValue(fileTokenManager.getToken(xcontext.getUserReference(), fileId, userCanWrite).toString());
```

Comment:
There's something I don't understand. You expose a REST API that **anyone** can call:

* some end-points **retrieve** data (``get`` and ``getContents``) so you need to be sure that the caller has view right
* some end-points **update** data (``postContents``) so you need to be sure that the caller has edit right

These end-points are called by Collabora which is not authenticated so you can't check access rights for the current user. Instead, you use a token, that is passed to all these end-points. But then in order to **trust** the passed token you need to be sure that access rights were checked when the token was generated. But I don't see any access rights checks in ``getToken``. Am I missing something? This means that anyone can call the ``getToken`` REST end-point saying "I can write" and then use this token to get the content of an attachment and even update it without having neither view nor edit right.

We need to either check access rights in ``getToken`` or, if Collabora calls this unauthenticated then find another way to pass the token (after we check access rights).

Reply:
Updated it. Now, when a token is requested, I check the rights (view and edit) of the current user, store them on the token and use this to validate requests. Also, these rights are updated in case they have been changed since creation